languages/python/sqlalchemy-oso/sqlalchemy_oso/roles2.py

```diff
@@ -0,0 +1,477 @@
+            assert arg.operator == "Isa"
+            assert len(arg.args) == 2
+            assert arg.args[0] == Variable("_this")
+            pattern = arg.args[1]
+            t = pattern.tag
+            name = result["bindings"]["name"]
+            permissions = result["bindings"]["permissions"]
+            role_defs = result["bindings"]["roles"]
+
+            python_class = (self.oso.host.classes[t],)
+
+            if isinstance(permissions, Variable):
+                permissions = set()
+            else:
+                permissions = set(permissions)
+
+            if isinstance(role_defs, Variable):
+                role_names = set()
+            else:
+                role_names = set(role_defs.keys())
+
+            resource = Resource(
+                python_class=python_class,
+                name=name,
+                actions=permissions,
+                roles=role_names,
+            )
+            self.resources[resource.name] = resource
+
+            permissions = [
+                Permission(name=action, python_class=python_class)
+                for action in permissions
+            ]
+            for permission in permissions:
+                self.permissions.append(permission)
+
+            role_definitions.append((python_class, role_defs))
+
+        for python_class, role_defs in role_definitions:
+            if not isinstance(role_defs, Variable):
+                for name, role_def in role_defs.items():
+                    role_permissions = []
+                    if "perms" in role_def:
+                        for permission in role_def["perms"]:
+                            if ":" in permission:
+                                resource_name, action = permission.split(":", 1)
+                                assert resource_name in self.resources
+                                permission_python_class = self.resources[
+                                    resource_name
+                                ].python_class
+                            else:
+                                action = permission
+                                permission_python_class = python_class
+                            role_permissions.append(
+                                Permission(
+                                    name=action, python_class=permission_python_class
+                                )
+                            )
+
+                    implied_roles = []
+                    if "implies" in role_def:
+                        implied_roles = role_def["implies"]
+
+                    role = Role(
+                        name=name,
+                        python_class=python_class,
+                        permissions=role_permissions,
+                        implied_roles=implied_roles,
+                    )
+                    self.roles[role.name] = role
+
+        # Sync static data to the database.
+        self.session.execute("delete from role_permissions")
```

Comment:
This means any time the application starts (and calls `roles.enable()`) the entire role data is deleted and updated.

We probably need a workflow that performs this configuration & syncing of the policy with the database as a script, right?

Before release?

Reply:
Before the next release we are going to need some way to update this data from the policy without doing a complete replacement and a without eliminating any dynamic customizations (if we add that capability)